src/app/campaign-list/list/campaign-list.component.ts

```diff
@@ -0,0 +1,138 @@
+import { Component, OnChanges, ChangeDetectionStrategy, Input } from '@angular/core';
+import { Router } from '@angular/router';
+import { Observable } from 'rxjs';
+import { Campaign, CampaignListService, Facets, CampaignParams } from '../campaign-list.service';
+
+@Component({
+  selector: 'grove-campaign-list',
+  template: `
+    <grove-campaign-filter
+      [params]="params" [facets]="facets" (campaignListParams)="routeToParams($event)">
+    </grove-campaign-filter>
+    <h2>Campaigns</h2>
+    <ul>
+      <li *ngFor="let campaign of campaigns$ | async">
+        <grove-campaign-card [campaign]="campaign"></grove-campaign-card>
+      </li>
+      <ng-container *ngIf="loading$ | async as loading">
+        <li *ngFor="let c of loading; let i = index">
+          <grove-campaign-card></grove-campaign-card>
+        </li>
+      </ng-container>
+    </ul>
+    <hr>
+    <prx-paging
+      [currentPage]="currentPage"
+      [totalPages]="totalPer | campaignListTotalPages"
+      (showPage)="routeToParams({page: $event})">
+    </prx-paging>
+    <div class="count-of-total">Showing {{count}} of {{total}}</div>
+  `,
+  styleUrls: ['campaign-list.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class CampaignListComponent implements OnChanges {
+  @Input() routedParams: CampaignParams;
+  campaigns$: Observable<Campaign[]> = this.campaignListService.loadedCampaigns;
+
+  constructor(private campaignListService: CampaignListService,
+              private router: Router) {}
+
+  ngOnChanges() {
+    this.campaignListService.loadCampaignList(this.routedParams);
+  }
+
+  get loading$(): Observable<boolean[]> {
+    return this.campaignListService.loading;
+  }
+
+  get currentPage(): number {
+    return this.campaignListService.params.page;
+  }
+
+  get count(): number {
+    return this.campaignListService.count;
+  }
+
+  get total(): number {
+    return this.campaignListService.total;
+  }
+
+  get totalPer(): {total: number, per: number} {
+    return {total: this.campaignListService.total, per: this.campaignListService.params.per};
+  }
+
+  get params(): CampaignParams {
+    return this.campaignListService.params;
+  }
+
+  get facets(): Facets {
+    return this.campaignListService.facets;
+  }
+
+  routeToParams(partialParams: CampaignParams) {
+    let { page, advertiser, podcast, status, type, text, representative } = partialParams;
+
+    // this function takes partial parameters (what changed)
+    // mix in the existing this.params unless property was explicitly set in partialParams
+    if (!partialParams.hasOwnProperty('page') && this.params.page) {
```

Comment:
Would this be any different than just checking `page === undefined` or something?

Reply:
I didn't love this, but the campaign-filter.component emits `{type: undefined, page: 1}` when the None option is selected from a select (because it doesn't have a value, which is what material recommends). Maybe not great naming, but partialParams won't contain the full route parameters, just what changed in the filters. So I needed a way to determine if the partialParams was meant to clear the value (hasOwnProperty) or if it should get the value from the full route `this.params`. Sometimes the negative logic also makes things less readable for me, so I think that's partly why this is confusing.